app/dao/template_categories_dao.py

```diff
@@ -0,0 +1,87 @@
+import uuid
+
+from flask import current_app
+
+from app import db
+from app.dao.dao_utils import transactional
+from app.models import Template, TemplateCategory
+
+
+@transactional
+def dao_create_template_category(template_category: TemplateCategory):
+    if template_category.id is None:
+        template_category.id = uuid.uuid4()
+    db.session.add(template_category)
+
+
+def dao_get_template_category_by_id(template_category_id) -> TemplateCategory:
+    return TemplateCategory.query.filter_by(id=template_category_id).one()
+
+
+def dao_get_template_category_by_template_id(template_id) -> TemplateCategory:
+    return Template.query.filter_by(id=template_id).one().category
+
+
+# TODO: Add filters: Select all template categories used by at least 1 sms/email template
+def dao_get_all_template_categories(template_type=None, hidden=None):
+    query = TemplateCategory.query
+
+    if template_type is not None:
+        query = query.join(Template).filter(Template.template_type == template_type)
+
+    if hidden is not None:
+        query = query.filter(TemplateCategory.hidden == hidden)
+
+    return query.all()
+
+
+@transactional
+def dao_update_template_category(template_category: TemplateCategory):
+    db.session.add(template_category)
+    db.session.commit()
+
+
+@transactional
+def dao_delete_template_category_by_id(template_category_id, cascade=False):
+    """
+    Deletes a `TemplateCategory`. By default, if the `TemplateCategory` is associated with any `Template`, it will not be deleted.
+    If the `cascade` option is specified then the category will be forcible removed:
+    1. The `Category` will be dissociated from templates that use it
+    2. The template is assigned a category matching the priority of the previous category
+    3. Finally, the `Category` will be deleted
+
+    Args:
+        template_category_id (str): The id of the template_category to delete
+        cascade (bool, optional): Specify whether to dissociate the category from templates that use it to force removal. Defaults to False.
+
+    Raises:
+        e: _description_
+    """
+    template_category = dao_get_template_category_by_id(template_category_id)
+    templates = Template.query.filter_by(template_category_id=template_category_id).all()
+
+    if not templates or cascade:
+        # When there are templates and we are cascading, we set the category to a default
+        # that matches the template's previous category's priority
+        if cascade:
+            for template in templates:
+                # Get the a default category that matches the previous priority of the template, based on template type
+                default_category_id = _get_default_category_id(
+                    template_category.sms_process_type
```

Comment:
so I am hesitant about this line, as we are going to have template.process_type be empty as a default..so there won't be a previous template. Can we just set the category to DefaultLow?